File: Permutations (Medium, #46)/Permutations.py
class Solution:
    def permute(self, nums: list[int]) -> list[list[int]]:
        result = []
        def recurse(new_perm, new_nums):
            nonlocal result
            if len(new_nums) == 0:
                result.append(new_perm)
                return
            
            for i in new_nums:
                current_perm = new_perm.copy()
                current_num = new_nums.copy()

                current_perm.append(i)
                current_num.remove(i)
                recurse(current_perm, current_num)
        
        recurse([], nums)
        return result

if __name__ == '__main__':
    sol = Solution()
    print(sol.permute([1,2,3]))

File: Permutations (Medium, #46)/Permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result {};
        recurse(result, {}, nums);
        return result;
    }

    void recurse(vector<vector<int>>& result, vector<int> current, vector<int>& new_nums) {
        int length = new_nums.size();
        if (length == 0) {
            result.push_back(current);
            return;
        }

        for (int i = 0; i < length; ++i) {
            vector<int> new_current = current;
            new_current.push_back(new_nums[i]);
            vector<int> current_nums = new_nums;
            current_nums.erase(current_nums.begin() + i);
            recurse(result, new_current, current_nums);
        }
    }
};

